Add multiple subscription mechanism

RaitoBezarius opened the pull request:
<!---
    Thank you for contributing to Aqua Security.
    Please don't remove the template.
-->

## Description
As per #144 ; this PR enables a `multiple_subscribe` decorator which waits for multiple events to be fired before the hook is executed.
It also supports multiple predicates for each event.

It passes the set of `events` directly to the hook.

I guess we can refine it a lot more to fit more use cases, but I didn't want to go overkill, I believe the greatest generalization would be full logic `@complex_subscribe(SomeEvent & (EventA | ~EventB))` for example, and we could feed predicates using a dict. (where `&` means AND, `|` means OR and `~` means NOT).

That sounds cool, but in practice, is there a need for it?

## Fixed Issues

Fixes #144.

## Contribution checklist
 - [x] I have read the Contributing Guidelines.
 - [x] The commits refer to an active issue in the repository.
 - [x] I have added automated testing to cover this case.

iYehuda commented on tests/core/test_subscribe.py:
```diff
@@ -26,25 +26,38 @@ def __init__(self, event):
+    # SmartHunter and OnceHunter should run only.
+    assert counter == 2
+    counter = 0
+
```
I think this code deserves its own test (i.e, "test_subscribe_many").
The current test is already overloaded with "subscribe" and "subscribe_once" testing, so you can split that too (please).

iYehuda commented on tests/core/test_subscribe.py:
```diff
@@ -26,25 +26,38 @@ def __init__(self, event):
-    
+
+    # testing the multiple subscription mechanism
+    handler.publish_event(OnceOnlyEvent())
```
Please add test case of subscriber being triggered twice.

iYehuda commented on src/core/events/handler.py:
```diff
@@ -44,6 +47,13 @@ def wrapper(hook):
 
         return wrapper
 
+    def multiple_subscribe(self, events, predicates=None):
```
Rename to `subscribe_many` to align with `subscribe_once`

iYehuda commented:
I'd like this PR to get another pair of eyes on it since it's a little complicated.
@itaysk can you review it?

RaitoBezarius commented:
I addressed some comments, but it's not passing tests anymore, I have to see why. I'm just worrying about the fact of splitting the test functions might have created some implicit global state that I don't see. :/

RaitoBezarius commented:
Also, it brings some question.

If I have some hunter depending on A & B

If I publish A, A, A, …, A, B, B, B, …, B

How many times should the hunter be called?

iYehuda commented:
> Also, it brings some question.
> 
> If I have some hunter depending on A & B
> 
> If I publish A, A, A, …, A, B, B, B, …, B
> 
> How many times should the hunter be called?

That's a good point.
Let's examine the use case with an example:
* Wer'e running on Azure (so AzureMetadataAPI event is published)
* There are 2 vulnerable kubelet APIs (so SecureKubeletEvent is published twice)
Assume there's a new vulnerability in azure that depends on the former combination, we would like to get it triggered for every kubelet found but we need only one metadata api as prerequisite.

It is possible to have a case which additional context should be included in triggering, so supplying a predicate for two events can be done to filter context.
In such case, you might have event A published twice and event B published twice as well, but a registered hunter C might need to have the second A and first B.
This way, you'll need to trigger C for each combination of A and B events and let it filter.

danielsagi commented:
Closing due to inactivity
